### PR TITLE
[zen-observable] Add zen observable esm entry point

### PR DESCRIPTION
**Pull request by BenoitZugmeyer**
zen-observable added a `'zen-observable/esm'` entry point, exporting all code (index and extra) as an ES module. This PR add support for it.

Contrary to importing `'zen-observable'`, it doesn't expose the global namespace `ZenObservable`.  Instead, I find it nicer to exports types directly.

Note: I'm reopening a PR because my [last PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44468) has been closed by the bot but I still think it's a valuable improvement. As stated by @typescript-bot : "please open a new PR if you'd like to continue with this change". So here it is!

---

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/zenparsing/zen-observable/blob/master/esm.js
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/zen-observable/esm.d.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+): Observable<[A, B, C, D, E]>;
+export function zip<A, B, C, D, E, F>(
+    a: ObservableLike<A>,
+    b: ObservableLike<B>,
+    c: ObservableLike<C>,
+    d: ObservableLike<D>,
+    e: ObservableLike<E>,
+    f: ObservableLike<F>,
+): Observable<[A, B, C, D, E, F]>;
+export function zip<T>(...observables: Array<ObservableLike<T>>): Observable<T[]>;
```
Comment (aicest) [inline types/zen-observable/esm.d.ts:139]: Maybe:
```typescript
import Observable from '.'

/**
 * @see https://github.com/zenparsing/zen-observable/blob/master/esm.js
 */
export default Observable
export { Observable }
```

Comment (BenoitZugmeyer) [reply types/zen-observable/esm.d.ts]: I agree, let's have something closer to the upstream package architecture. I had to make a few changes to do this, let me know what you think.

**Review by aicest — APPROVED** [review]

**Comment by andrewbranch** [comment]
I might be misunderstanding the intent since I’m not familiar with the source library at all, but I’m not confident this does what you intend. Because the `zen-observable/esm` entrypoint imports the global-augmenting module, the global namespace is still augmented:

image

**Comment by BenoitZugmeyer** [comment]
> I might be misunderstanding the intent since I’m not familiar with the source library at all, but I’m not confident this does what you intend. Because the `zen-observable/esm` entrypoint imports the global-augmenting module, the global namespace is still augmented:
> 
> image

You are right, the global namespace is still defined when importing `esm`. This is a side effect of [the change](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46628/commits/82b5d2e10bb3e28d2b57e3d5fde09e8a804719e8) I did to comply to the @aicest [review](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46628#discussion_r467470226).

Maybe we could have a `./lib/index.d.ts` file regrouping all definitions (at the exception of the namespace), and `./esm.d.ts`, `./index.d.ts` and `./extras.d.ts` exporting the correct types for the three entry-points exposed by the library (see [index](https://github.com/zenparsing/zen-observable/blob/master/index.js) [esm](https://github.com/zenparsing/zen-observable/blob/master/esm.js) [extras](https://github.com/zenparsing/zen-observable/blob/master/extras.js)). The `ZenObservable` namespace could be defined only in `./index.d.ts`. WDYT?

**Comment by andrewbranch** [comment]
You typically want to avoid adding files in the type definitions that don’t correspond to files in the source package, since that module path will appear to be importable to the consumer. I was thinking the way you have it now could possibly just be switched—if `esm.d.ts` declared all the non-globals, perhaps `index.d.ts` could import it and expose the types to the global namespace?

**Comment by BenoitZugmeyer** [comment]
> `esm.d.ts` declared all the non-globals, perhaps `index.d.ts` could import it and expose the types to the global namespace?

That's exactly what I did before the review: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46628/commits/365e50e56b3a4fb0f6a2f71319fde095a342375f . Should I revert my last commit? @aicest what's your opinion on this?

**Comment by BenoitZugmeyer** [comment]
Ready to merge

**Pull request merged**